### PR TITLE
Add support for tokio tracing for actix Service.

### PR DESCRIPTION
**Pull request by avranju**

### CODE REVIEW

**Comment by JohnTitor** [comment]
I'm not sure this should be in actix-net, not third-party crate. CC @fafhrd91

**Comment by fafhrd91** [comment]
I asked Raj to add it to actix-net

File: actix-tracing/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+
+use actix_service::{apply, IntoServiceFactory, Service, ServiceFactory, Transform};
+use futures_util::future::{ok, Either, Ready};
+use tracing_futures::{Instrument, Instrumented};
+
+/// Trait representing a type that is able to produce spans for a given value.
+///
+/// The generic argument `T` is typically the request type of a `Service`
+/// implementation.
+pub trait MakeSpan<T> {
```
Comment (fafhrd91) [inline actix-tracing/src/lib.rs]: what is the reason for trait here? why not just simple `Fn(&T) -> Option<Span>`?

File: actix-tracing/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    /// Creates a new span for the given request value.
+    ///
+    /// By default, this function returns `None`.
+    fn span_for(&self, _req: &T) -> Option<tracing::Span> {
```
Comment (fafhrd91) [inline actix-tracing/src/lib.rs]: should be required method, no reason for default impl

File: actix-tracing/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+///     web_service,
+///     |req: &Request| Some(span!(Level::INFO, "request", req.id))
+/// );
+/// ```
+///
+/// [`MakeSpan`]: trait.MakeSpan.html
+pub fn trace<S, U, T>(
+    service_factory: U,
+    make_span: T,
+) -> impl ServiceFactory<
```
Comment (fafhrd91) [inline actix-tracing/src/lib.rs]: it won't be possible to use `+Clone`  with `impl Trait` here

File: actix-tracing/src/lib.rs (hunk 1 above)
Comment (avranju) [reply actix-tracing/src/lib.rs]: Made it a trait in the anticipation that we might have more span making functions in the future (for `poll_ready` for instance) and provided a blanket impl for `Fn(&T) -> Option<Span>`. But I don't feel strongly about this. We can always have a trait later if we think its appropriate. Will change.

File: actix-tracing/src/lib.rs (hunk 2 above)
Comment (avranju) [reply actix-tracing/src/lib.rs]: Agree.

File: actix-tracing/src/lib.rs (hunk 3 above)
Comment (avranju) [reply actix-tracing/src/lib.rs]: Not sure I understand this. I was able to add a `Clone` bound on the `Fn`.

Comment (fafhrd91) [reply actix-tracing/src/lib.rs]: Resulted factory wouldn’t have clone even if input service factory would have clone

Comment (fafhrd91) [reply actix-tracing/src/lib.rs]: Let’s return `ApplyTransform` type instead of impl ServiceFactory

Comment (avranju) [reply actix-tracing/src/lib.rs]: Changed the return type to `ApplyTransform`.

Still don't understand the `Clone` comment. The `Clone` bound is on the closure and not on the factory returned from `trace`.

Comment (fafhrd91) [reply actix-tracing/src/lib.rs]: but if you want to make clonable service factory, it would be not possible with impl Trait

**Comment by fafhrd91** [comment]
thanks!

**Comment by cdbattags** [comment]
Any documentation yet for using this with `actix-web`?

**Pull request merged**